BOM-1195: Upgrade django-oscar to 2.x

iamsobanjaved opened the pull request:
**JIRA:** [BOM-1195](https://openedx.atlassian.net/browse/BOM-1195)

- Upgrade django-oscar to 2.x

### Changes

- Removed OSCAR_MAIN_TEMPLATE_DIR and added 'oscar/' prefix to oscar's templates
- Removed Oscar's get_core_apps and replaced with manually entering core apps in _oscar.py
- added orders argument to get_hourly_report

**Forked apps:**
- changed filename from config.py to apps.py
- updated init.py files to change default_app_config
- moved content of app.py to apps.py into OscarConfig's subclass
- removed application variable from app.py
- urls are now accessed through Django's apps.get_app_config()
- ran migrations for oscar's latest changes

**Reference**
https://django-oscar.readthedocs.io/en/latest/releases/v2.0.html

Ayub-Khan commented on ecommerce/extensions/utils.py:
```diff
@@ -0,0 +1,7 @@
+
+def exclude_app_urls(urls, app_name):
```
A Doc string for method and file would be nice.

Ayub-Khan commented on ecommerce/settings/base.py:
```diff
@@ -200,7 +198,7 @@
                 'django.contrib.messages.context_processors.messages',
                 'django.template.context_processors.request',
                 'oscar.apps.search.context_processors.search_form',
-                'oscar.apps.promotions.context_processors.promotions',
```
Add a detailed comment about the change.

Ayub-Khan commented on ecommerce/tests/mixins.py:
```diff
@@ -132,7 +132,7 @@ def set_jwt_cookie(self, system_wide_role=SYSTEM_ENTERPRISE_ADMIN_ROLE, context=
 
 class BasketCreationMixin(UserMixin, JwtMixin):
     """Provides utility methods for creating baskets in test cases."""
-    PATH = reverse('api:v2:baskets:create')
```
Why reverse_lazy

Ayub-Khan commented on requirements/base.in:
```diff
@@ -24,7 +24,7 @@ drf-extensions==0.3.1
 edx-django-release-util
 edx-django-utils
-edx-drf-extensions
+edx-drf-extensions<2.4.7		# causing failures. Upgrade in separate PR
```
do this using the pins.txt

Ayub-Khan replied:
'The latest versions are not supported yet. '

Ayub-Khan commented:
Good Work.

iamsobanjaved commented:
@jmbowman Kindly have a look at it and do tag some related persons. PR is ready except for few quality failures which will be resolved by tomorrow.

jmbowman commented:
I've started looking at this and asked for some help with it, but probably won't have feedback ready until sometime tomorrow.

robrap commented on ecommerce/templates/oscar/dashboard/orders/order_list.html:
```diff
@@ -45,7 +45,7 @@ <h1>{% trans "Orders" as tmsg %}{{ tmsg | force_escape }}</h1>
       <form action="" method="post" class="order_table" id="orders_form">
         {% csrf_token %}
-        {% include "dashboard/orders/partials/bulk_edit_form.html" with status=active_status %}
+        {% include "oscar/dashboard/orders/partials/bulk_edit_form.html" with status=active_status %}
```
Should all of these be updated?  I just did a straight search for `include "dashboard/`, and found a bunch more.

robrap replied:
I'm marking this as resolved.  Thanks.

robrap commented:
FYI: Here is a link to the migration SQL: https://gist.github.com/robrap/310b0b2d414c3dcb99a01825c0023564

awais786 commented:
@robrap I also generated the link but i was verifying it
https://gist.github.com/awais786/dcbd3fe05a3993a1a8ab9e458c55de6f

robrap commented:
My apologies for stepping on your toes @awais786.  I know it was your team's action item, but I also know you were working on a lot, and I didn't see it yet, and hoped you were alseep. ;)  I guess not.

awais786 commented:
> My apologies for stepping on your toes @awais786. I know it was your team's action item, but I also know you were working on a lot, and I didn't see it yet, and hoped you were alseep. ;) I guess not.

Thanks for the support.

iamsobanjaved commented:
@robrap We have changed ecommerce/extensions/dashboard/refunds/apps.py 

 to this:

    class RefundsDashboardConfig(OscarDashboardConfig):

    label = 'refunds_dashboard'

    name = 'ecommerce.extensions.dashboard.refunds'

    verbose_name = _('Refunds Dashboard')

    default_permissions = ['is_staff', ]

    permissions_map = {
        'refunds-list': (['is_staff'], ['partner.dashboard_access']),
        'refunds-detail': (['is_staff'], ['partner.dashboard_access']),
    }

    # pylint: disable=attribute-defined-outside-init
    def ready(self):
        super().ready()
        self.refund_list_view = get_class('dashboard.refunds.views', 'RefundListView')
        self.refund_detail_view = get_class('dashboard.refunds.views', 'RefundDetailView')

    def get_urls(self):
        urls = [
            url(r'^$', self.refund_list_view.as_view(), name='refunds-list'),
            url(r'^(?P<pk>[\d]+)/$', self.refund_detail_view.as_view(), name='refunds-detail'),
        ]
        return self.post_process_urls(urls)

imitating 
https://github.com/django-oscar/django-oscar/blob/2.0.0/src/oscar/apps/dashboard/orders/apps.py

and changed `dashboard:refunds:detail` to `dashboard:refunds-detail` and `dashboard:refunds:list` to `dashboard:refunds-list`. We have not pushed these changes because it will break PR but this is not working on local. You may guide us in this regard

iamsobanjaved commented on ecommerce/settings/base.py:
```diff
@@ -200,7 +196,12 @@
+                # Hence, it is removed from ecommerce as not found any use of this app in ecommerce. Your valuable
+                # opinion is appreciated
+
+                # 'oscar.apps.promotions.context_processors.promotions',
```
Deleted

iamsobanjaved commented on ecommerce/extensions/config.py:
```diff
@@ -0,0 +1,27 @@
+    # URLs are only visible to users with staff permissions
+    default_permissions = 'is_staff'
+
+    # TODO -- I will remove it later.
```
Cleaned

iamsobanjaved commented on ecommerce/extensions/dashboard/apps.py:
```diff
@@ -0,0 +1,42 @@
+            url(r'^reports/', self.reports_app.urls),
+            url(r'^orders/', self.orders_app.urls),
+            url(r'^users/', self.users_app.urls),
+            # url(r'^content-blocks/', self.promotions_app.urls[0)),
```
Deleted

iamsobanjaved commented on ecommerce/settings/_oscar.py:
```diff
@@ -15,25 +14,48 @@
+    'ecommerce.extensions.dashboard.refunds',
     'ecommerce.extensions.dashboard.orders',
     'ecommerce.extensions.dashboard.users',
     'ecommerce.extensions.offer',
```
1. We can't leave out both, because we are updating about urls in their config.
2. Sorted according to given order.

robrap replied:
Great.  Thanks.

iamsobanjaved commented on ecommerce/extensions/partner/admin.py:
```diff
@@ -4,7 +4,7 @@
 from oscar.core.loading import get_class
 
-Catalog = get_class('ecommerce.extensions.catalogue.models', 'Catalog')
+Catalog = get_class('catalogue.models', 'Catalog')
```
It worked with the previous version of oscar, we could import that way too. But with the newer version of oscar, we can't.

robrap replied:
Thanks.

robrap commented:
Nice job.  Thank you.  I hope everything goes smoothly on Stage on Monday.  Good luck.
